File: cpp/leetcode/022.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        dfs(n, n, "", res);
        return res;
    }
    void dfs(int l, int r, string ans, vector<string> &res) {
        if (l == 0 && r == 0) {
            res.push_back(ans);
        }
        if (l > 0 && l <= r) {
            dfs(l - 1, r, ans + "(", res);
        }
        if (r > 0 && l <= r) {
            dfs(l, r - 1, ans + ")", res);
        }
    }
};

File: python/leetcode/022.py
class Solution(object):
    def generateParenthesis(self, n):
        res = []
        self.dfs(n, n, "", res)
        return res

    def dfs(self, l, r, ans, res):
        if l == 0 and r == 0:
            res.append(ans)
        if l > 0 and l <= r:
            self.dfs(l - 1, r, ans + '(', res)
        if r > 0 and l <= r:
            self.dfs(l, r - 1, ans + ')', res)